File: Knou/2023년 2학기/simulation/java_src(file)/EX3_1.java
/*--------------------------------------------------------------
 * 복수창구 대기행렬 시뮬레이션
 * tpump    : 고객에 대한 봉사시간(단위 : 분)
 * queue    : 줄에서 기다리고 있는 사람(차량) 수, 대기행렬의 길이
 * prarr    : 고객이 1분내에 도착할 확률
 * arrive   : 0 = 고객이 도착하지 않은 경우,
 *            1 = 고객이 도착한 경우
 * time     : 시뮬레이션의 현재시간(단위 : 분)
 * tstep    : 시뮬레이션 진행 단위 시간
 * tlimit   : 총 시뮬레이션 수행 시간
 * totarr   : 도착한 총 고객수
 * totque   : 대기행렬 queue 의 전체 합 (＝전체 고객의 총 대기시간)
 * aveque   : 대기행렬 queue 의 평균길이
 * seed     : 난수의 초기값
 * tpump(i) : 봉사시간
 *-------------------------------------------------------------*/
import java.lang.*;
import java.io.*;

class DoubleQueueing {
        final static int SEED = 35213;
        public int n, p, seed;
        public float up, mean;
        public String StrTime = new String();

        public DoubleQueueing(){
               mean = 4;
               n = SEED;
               seed = SEED;
        }

        public int random(int np, float u){
                np = np * 843314861 + 453816693;
                if(np < 0){
                        np = np + 2147483647;
                        np = np + 1;
                }
                up = (float)(np * 0.4656612e-9);
                return np;
        }

        public int poissn(int np, int pp){
                float b, prod;
                pp = 0;
                b = (float)(Math.exp(-mean));

                prod = 1;
                n = random(np, up);
                prod = prod * up;

                while(prod >= b){
                        n = random(n, up);
                        prod = prod * up;
                        ++pp;
                }
                return pp;
        }

        public void ConvDataToString(int time){
               int i, len;

               StrTime = "";
               StrTime = StrTime.valueOf(time);
               len = StrTime.length();
               if(len < 3)
                    for(i = 0; i < (3-len); i++)
                         StrTime = "0" + StrTime;
        }

        public void ComputeQueueing(){
                int i, queue = 0, totque = 0, totarr=0, arrive, tstep = 1;
                float prarr=1.0f/3.0f, time=0, tlimit=100, aveque=0;
                float tpump[] = new float[3];
                tpump[1] = 0;
                tpump[2] = 0;
                PrintWriter out=null;

                try {
                     File f=new File("C:\\WORK\\EX3_1.OUT");
                     FileWriter fw=new FileWriter(f) ;
                     BufferedWriter bw=new BufferedWriter(fw);
                     out=new PrintWriter(bw);

                     out.println(" SIMULATION FOR DOUBLE QUEUEING SYSTEM");
                     out.println("=========================================");
                     out.println("THE TIME STEP           = " + tstep);
                     out.println("THE TIME LIMIT          = " + (int)tlimit);
                     out.println("THE ARRIVAL PROBABILITY = " + (float)((int)(prarr*100))/100);
                     out.println("THE POISSON MEAN        = " + (int)mean);
                     out.println("THE SEED                = " + n);
                     out.println("-----------------------------------------");
                     out.println(" TIME  ARRIVAL  QUEUE  TPUMP[1]  TPUMP[2]");

                     while(time < tlimit) {
                             time = time + tstep;
                             arrive = 0;
                             seed = random(seed, up);
                             if(up < (prarr*tstep)) {
                                     arrive = 1;
                                     queue = queue + arrive;
                                     totarr = totarr + 1;
                             }
                             i = 0;

                             while(i < 2) {
                                     i++;
                                     if(tpump[i] > 0) {
                                             tpump[i] = tpump[i]-tstep;
                                             if(tpump[i] < 0.0)
                                                      tpump[i] = 0.0f;
                                     }
                                     if((tpump[i]==0.0) && (queue!=0)) {
                                             queue = queue - 1;
                                             p = poissn(n, p);
                                             tpump[i] = p;
                                     }
                             }
                             totque = totque + queue;
                             ConvDataToString((int)time);
                             out.println(" " + StrTime + "      " + arrive + "       " + queue + "       " +
                                                (int)tpump[1] + "         " + (int)tpump[2]);
                     }
                     aveque = totque / (tlimit*tstep);

                     out.println("-------------------------------------");
                     out.println("MEAN QUEUE LENGTH = " + aveque);
                     out.println("THE TOTAL ARRIVALS = " + totarr);
                     out.println("-------------------------------------");
                 } catch(IOException ioe) { }
                   finally {
                         if (out != null) out.close();
                   }
        }
}

public class EX3_1 {
        public static void main(String[] args){
                DoubleQueueing g = new DoubleQueueing();
                g.ComputeQueueing();
        }
}

File: Knou/2023년 2학기/simulation/VC++2019/ex31/ex3-1.c
// SIMULATION FOR DOUBLE QUEUEING SYSTEM                        //
// TO EVALUATE MEAN QUEUE LENGTH                                //
// Two server and                                               //
// The service time has poisson distribution.                   //

// TLIMIT   : the time limit expressed in minutes               //
// TIME     : the time expressed in minutes                     //
// TSTEP    : the time increment in minutes                     //
// TPUMP(i) : the service time that a car has left in server i  //
// PRARR    : the probability of a car arriving in any minute   //
// AVEQUE   : the average queue length                          //
// TOTQUE   : the totals of queue length                        //
// TOTARR   : the totals of arrivals                            //

// PROGRAM  EX3-1                                               //

#include<stdio.h>
#include<math.h>

#define SEED 35213;

long p;
float mean = 4, b;


//  RANDOM NUMBER GENERATOR  //
void random(long *np, float *up)
{
       *np = *np * 843314861 + 453816693;
       if(*np < 0)
	      { *np = *np +2147483647;
		*np = *np + 1;
	      }
       *up = *np * 0.4656612e-9;
}


//  THE SERVICE TIME HAS POISSON DISTRIBUTION  //
void poissn(long *np, int *pp)
{
       float prod,b,u;

       *pp = 0;
       b = exp(-mean);
       prod = 1;
       random(np,&u);
       prod=prod*u;

       while (prod >= b)
	     { random(np,&u);
	       prod = prod * u;
	       ++(*pp);
	     }
}


//  MAIN PROGRAM  //
void main()
{
       long seed,nseed;
       int  p, queue=0, totque=0, totarr=0, arrive, tstep=1, i;
       float prarr=1.0/3.0, tpump[3], time=0.0, tlimit=100, mean=4, u=0, aveque=0;
       FILE *output;
       output=fopen("c:\\work\\ex3-1.out","wt");

	   seed=SEED;
	   nseed = SEED;
	   tpump[1] = 0;
	   tpump[2] = 0;

       fprintf(output,"        SIMULATION FOR DOUBLE QUEUEING SYSTEM       \n\n");
       fprintf(output,"THE TIME STEP            =  %d \n",tstep);
       fprintf(output,"THE TIME LIMIT           =  %3.0f \n",tlimit);
       fprintf(output,"THE ARRIVAL PROBABILITY  =  %4.2f \n",prarr);
       fprintf(output,"THE POISSON MEAN         =  %1.0f \n",mean);
       fprintf(output,"THE SEED                 =  %ld \n\n",seed);
       fprintf(output,"     TIME     ARRIVAL     QUEUE     TPUMP[1]     TPUMP[2] \n");

       while (time < tlimit)
	       {
		 time = time + tstep;
		 arrive = 0;
		 random(&seed, &u);
		 if (u < (prarr*tstep))
			{ arrive = 1;
			  queue = queue + arrive;
			  totarr = totarr + 1;
			}
		 i = 0;
			while ( i < 2 )
			    { i = i + 1;
			      if (tpump[i] > 0)
				    { tpump[i] = tpump[i] - tstep;
			      if (tpump[i] < 0.0) tpump[i] = 0.0;
				    }
			      if ((tpump[i] == 0) && (queue != 0))
				    { queue = queue - 1;
				      poissn(&nseed, &p);
				      tpump[i] = p;
				    }
			    }   //  END OF INNER WHILE  //
		 totque = totque + queue;
		 fprintf(output,"     %3.0f         %d          %d          %2.0f           %2.0f \n"
		 ,time,arrive,queue,tpump[1],tpump[2]);
	       }   //  END OF  OUTER WHILE  //
       aveque = totque / (tlimit/tstep);
       fprintf(output,"  MEAN QUEUE LENGTH  =  %4.2f \n",aveque);
       fprintf(output,"  THE TOTAL ARRIVALS =  %d \n",totarr);
       fprintf(output,"  THE TOTAL QUEUE =  %d \n",totque);
       fclose(output);
}
